src/opt/horizontal_fusion_impl.cpp

```diff
@@ -0,0 +1,839 @@
+            for(auto&& input : hash_inputs[hash_id])
+            {
+                std::vector<instruction_ref> instrs = get_instrs(input->id);
+                if(instrs.size() != 1)
+                {
+                    doit = false;
+                    break;
+                }
+                visited[instrs.at(0)] = true;
+            }
+            if(!doit)
+                continue;
+
+            std::vector<instruction_ref> base_instrs = get_instrs(hash_id);
+            instruction_ref ins0                     = base_instrs.at(0);
+
+            // save original dimensions.
+            std::vector<std::vector<std::size_t>> lens;
+            std::vector<int> clusters;
+            int enum_ndx = 0;
+            for(auto&& ins : base_instrs)
+            {
+                lens.push_back(ins->get_shape().lens());
+                for(unsigned long i = 0; i < ins->outputs().size(); i++)
+                    clusters.push_back(enum_ndx);
+                enum_ndx++;
+            }
+            orig_dims[ins0]     = lens;
+            orig_clusters[ins0] = clusters;
+
+            if(ins0->inputs().size() == 1)
+            {
+                // concat single input instructions.
+                instruction_ref input = ins0->inputs().at(0);
+                if(split_axis.find(input) != split_axis.end())
+                {
+                    ins0->set_shape(input->get_shape());
+                    remove_redundant_roots(base_instrs);
+                    update_hash_tree(hash_id);
+                    last_hash_id     = hash_id;
+                    split_axis[ins0] = split_axis[input];
+                }
+                continue;
+            }
+
+            std::vector<std::vector<instruction_ref>> all_inputs;
+            int axis = -1;
+
+            if(!collect_inputs(all_inputs, axis, base_instrs, visited, root, split_axis))
+                continue;
+
+            transform_layers(all_inputs, root, axis, base_instrs);
+            update_hash_tree(hash_id);
+            last_hash_id = hash_id;
+        }
+
+        if(last_hash_id != 0)
+            transform_output(last_hash_id, split_axis, orig_dims, orig_clusters);
+    }
+}
+
+std::vector<instruction_ref>
+horizontal_fusion_impl::walk(instruction_ref ins,
+                             std::unordered_map<instruction_ref, bool>& visited)
+{
+
+    std::stack<instruction_ref> stk;
+    for(auto&& input : ins->inputs())
+    {
+        if(visited.find(input) == visited.end())
+            stk.push(input);
+    }
+
+    std::vector<instruction_ref> ret;
+    while(!stk.empty())
+    {
+        instruction_ref top = stk.top();
+        if((top->inputs().size() > 1) || (top->outputs().size() > 1) ||
+           (top->inputs().empty() && (top->name() != "@literal")))
+        {
+            ret.clear();
+            return ret;
+        }
+        else if(top->inputs().empty() || (visited.find(top) != visited.end()))
```

Comment:
This seems like it is only true if is `@literal`. Can you add more of an explanation why you are traversing like this? I am not really following all of these conditionals here and why they are needed.

Reply:
comments added.